### DISCUSSION TITLE
Tutorial to control the Create 3 robot using ROS 2

### DISCUSSION THREAD

**Discussion by yorambenhaim** [iRobotEducation/create3_docs#412] [Questions]
### How are you connecting to your Create 3?

Wi-Fi (ROS 2)

### Computer(s) Model(s) and Operating System(s)

Lenovo laptop (Intel i7) , Ubuntu 22.04 

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

G4.5

### Which RMW is your robot running?

FastRTPS

### Does your robot have an assigned namespace? If so, please share.

No

### Is the robot connected to a network? If so, what is the network type?

WIFI 2.4GHZ

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

None

### What is the Adapter Board's USB/BLE Toggle currently switched to?

Bluetooth (default)

### Describe your question.

Hello friends , a beginner's questions :(
Im a SW engineer, I got a brand-new robot and start playing with it 
I aim to have the ability to program the iRobot (Python or CPP) through ROS2/WIFI. i.e. Single Host/Single Robot. 

I was able to gain initial control of the robot using a few cli commands over my Home WIFI net . For example : 
_**ros2 topic pub -r 20 /cmd_vel geometry_msgs/msg/Twist "{linear: {x: 0.2, y: 0.0, z: 0.0}, angular: {x: 0.0, y: 0.0, z: 0.0}}"**_

My questions : 
1. Why I don't have the path 'irobot_create_msgs' for message-type ?? Note: Most examples use this path
2. Can you recommend a tutorial guide for software developing (PY/CPP) to have a HOST/ROBOT movement control at first?
    Note: Without BT , Using only the HUMBLE ROS2 (Over 2.4GHZ WIFI) for communication between the laptop computer to the Robot

  Thank you in advanced.
  Yoram Ben-Haim

**Comment by alsora** [comment]
Hi @yorambenhaim 

> Why I don't have the path 'irobot_create_msgs' for message-type ?? Note: Most examples use this path

Did you install the package? `sudo apt-get update && sudo apt-get install ros-humble-irobot-create-msgs`

> Can you recommend a tutorial guide for software developing (PY/CPP) to have a HOST/ROBOT movement control at first?

We have developed a tutorial for ROS 2 beginners that will guide you from the basic concept and CLI control to write simple python applications. 
You can find the lessons here https://edu.irobot.com/units/getting-started-with-ros-2
The main advantage of this tutorial is that it's focused on Create 3.

You can also find Python and C++ examples in the official ROS 2 tutorial https://docs.ros.org/en/humble/Tutorials.html
In particular, have a look at the "beginners: client libraries" section and the "intermediate" section

**Reply by yorambenhaim** [reply]
Thank you alsora 
I already tried the **_sudo apt-get update && sudo apt-get install ros-humble-irobot-create-msgs_** command

And having these failures : 

[sudo] password for administrator: 
Ign:1 http://il.archive.ubuntu.com/ubuntu jammy InRelease                                                                                                                                                         
Ign:2 http://security.ubuntu.com/ubuntu jammy-security InRelease                                                                                                                                                  
Ign:3 https://packages.microsoft.com/repos/vscode stable InRelease                                                                                                                                                
Ign:4 https://download.docker.com/linux/ubuntu jammy InRelease                                                                                                                                                    
Ign:5 https://dl.google.com/linux/chrome/deb stable InRelease                                                 
Ign:6 http://packages.ros.org/ros2/ubuntu jammy InRelease                                                     
Ign:7 http://il.archive.ubuntu.com/ubuntu jammy-updates InRelease                                             
Ign:8 http://il.archive.ubuntu.com/ubuntu jammy-backports InRelease
Ign:5 https://dl.google.com/linux/chrome/deb stable InRelease
Ign:4 https://download.docker.com/linux/ubuntu jammy InRelease
Ign:3 https://packages.microsoft.com/repos/vscode stable InRelease
Ign:1 http://il.archive.ubuntu.com/ubuntu jammy InRelease
Ign:2 http://security.ubuntu.com/ubuntu jammy-security InRelease
Ign:6 http://packages.ros.org/ros2/ubuntu jammy InRelease
Ign:7 http://il.archive.ubuntu.com/ubuntu jammy-updates InRelease
Ign:8 http://il.archive.ubuntu.com/ubuntu jammy-backports InRelease
Ign:4 https://download.docker.com/linux/ubuntu jammy InRelease
Ign:5 https://dl.google.com/linux/chrome/deb stable InRelease
Ign:3 https://packages.microsoft.com/repos/vscode stable InRelease
Ign:1 http://il.archive.ubuntu.com/ubuntu jammy InRelease
Ign:2 http://security.ubuntu.com/ubuntu jammy-security InRelease
Ign:6 http://packages.ros.org/ros2/ubuntu jammy InRelease
Ign:7 http://il.archive.ubuntu.com/ubuntu jammy-updates InRelease
Ign:8 http://il.archive.ubuntu.com/ubuntu jammy-backports InRelease
Err:5 https://dl.google.com/linux/chrome/deb stable InRelease
  Could not resolve 'proxy-iil.intel.com'
Err:4 https://download.docker.com/linux/ubuntu jammy InRelease
  Could not resolve 'proxy-iil.intel.com'
Err:3 https://packages.microsoft.com/repos/vscode stable InRelease
  Could not resolve 'proxy-iil.intel.com'
Err:2 http://security.ubuntu.com/ubuntu jammy-security InRelease
  Could not resolve 'proxy-iil.intel.com'
Err:1 http://il.archive.ubuntu.com/ubuntu jammy InRelease
  Could not resolve 'proxy-iil.intel.com'
Err:6 http://packages.ros.org/ros2/ubuntu jammy InRelease
  Could not resolve 'proxy-iil.intel.com'
Err:7 http://il.archive.ubuntu.com/ubuntu jammy-updates InRelease
  Could not resolve 'proxy-iil.intel.com'
Err:8 http://il.archive.ubuntu.com/ubuntu jammy-backports InRelease
  Could not resolve 'proxy-iil.intel.com'
Reading package lists... Done
W: Failed to fetch http://il.archive.ubuntu.com/ubuntu/dists/jammy/InRelease  Could not resolve 'proxy-iil.intel.com'
W: Failed to fetch http://il.archive.ubuntu.com/ubuntu/dists/jammy-updates/InRelease  Could not resolve 'proxy-iil.intel.com'
W: Failed to fetch http://il.archive.ubuntu.com/ubuntu/dists/jammy-backports/InRelease  Could not resolve 'proxy-iil.intel.com'
W: Failed to fetch http://security.ubuntu.com/ubuntu/dists/jammy-security/InRelease  Could not resolve 'proxy-iil.intel.com'
W: Failed to fetch https://download.docker.com/linux/ubuntu/dists/jammy/InRelease  Could not resolve 'proxy-iil.intel.com'
W: Failed to fetch https://dl.google.com/linux/chrome/deb/dists/stable/InRelease  Could not resolve 'proxy-iil.intel.com'
W: Failed to fetch http://packages.ros.org/ros2/ubuntu/dists/jammy/InRelease  Could not resolve 'proxy-iil.intel.com'
W: Failed to fetch https://packages.microsoft.com/repos/vscode/dists/stable/InRelease  Could not resolve 'proxy-iil.intel.com'
W: Some index files failed to download. They have been ignored, or old ones used instead.
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  ros-humble-irobot-create-msgs
0 upgraded, 1 newly installed, 0 to remove and 242 not upgraded.
Need to get 475 kB of archives.
After this operation, 9,973 kB of additional disk space will be used.
Ign:1 http://packages.ros.org/ros2/ubuntu jammy/main amd64 ros-humble-irobot-create-msgs amd64 2.1.0-1jammy.20230112.153759
Ign:1 http://packages.ros.org/ros2/ubuntu jammy/main amd64 ros-humble-irobot-create-msgs amd64 2.1.0-1jammy.20230112.153759
Ign:1 http://packages.ros.org/ros2/ubuntu jammy/main amd64 ros-humble-irobot-create-msgs amd64 2.1.0-1jammy.20230112.153759
Err:1 http://packages.ros.org/ros2/ubuntu jammy/main amd64 ros-humble-irobot-create-msgs amd64 2.1.0-1jammy.20230112.153759
  Could not resolve 'proxy-iil.intel.com'
E: Failed to fetch http://packages.ros.org/ros2/ubuntu/pool/main/r/ros-humble-irobot-create-msgs/ros-humble-irobot-create-msgs_2.1.0-1jammy.20230112.153759_amd64.deb  Could not resolve 'proxy-iil.intel.com'
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?

**Reply by alsora** [reply]
It looks like your ubuntu system is not correctly configured.
This error seems unrelated to ROS 2 or create 3.

I'm not familiar with this problem and a quick Google search didn't show anything.

**Comment by yorambenhaim** [comment]
> Hi @yorambenhaim
> 
> > Can you recommend a tutorial guide for software developing (PY/CPP) to have a HOST/ROBOT movement control at first?
> 
> We have developed a tutorial for ROS 2 beginners that will guide you from the basic concept and CLI control to write simple python applications. You can find the lessons here https://edu.irobot.com/units/getting-started-with-ros-2 The main advantage of this tutorial is that it's focused on Create 3.
> 
> You can also find Python and C++ examples in the official ROS 2 tutorial https://docs.ros.org/en/humble/Tutorials.html In particular, have a look at the "beginners: client libraries" section and the "intermediate" section
~~~~
Thank you very much.
I looked at getting-started-with-ros-2 and they refer to ROS2-Galactic as a pre-requisite while I already installed the ROS2-Humble. Is it an issue ? Do you recommend to down-grade to Galactic ?  
B.R. 
Yoram Ben-Haim

**Reply by alsora** [reply]
It's strongly recommended to make sure that laptop and robot are using the same ROS 2 version (e.g. galactic or humble).
Note that even if the tutorial refer to Galactic, there will be almost no differences if using Humble (the main/only difference that comes to my mind is that we renamed some dock related topics/actions when moving from Galactic to Humble).

You should either update your robot to Humble or downgrade your laptop to Galactic or run a Docker container with ROS 2 Galactic.
